### PR TITLE
Add save, evaluate and predict functionality for example

### PR DESCRIPTION
**Pull request by nyghtowl**
Add save model, evaluate test data and who small sample of predictions for example. Also, included string labels for predict in order to print to screen.

Ran this locally to confirm runs before PR.

### CODE REVIEW

File: examples/cifar10_cnn.py [hunk 1]
```diff
@@ -15,10 +15,20 @@
...
+import pickle
+from IPython.utils.path import ensure_dir_exists
+import numpy as np
+
 batch_size = 32
 num_classes = 10
 epochs = 200
 data_augmentation = True
+num_pred_print = 20
+save_dir = os.getcwd() + "/saved_models/"
```
Comment (fchollet) [inline examples/cifar10_cnn.py]: Throughout this file, `"` is mixed up with `'`. Only use `'`, for consistency.

File: examples/cifar10_cnn.py [hunk 2]
```diff
@@ -15,10 +15,20 @@
 from keras.layers import Dense, Dropout, Activation, Flatten
 from keras.layers import Conv2D, MaxPooling2D
 
+import os
+import pickle
+from IPython.utils.path import ensure_dir_exists
```
Comment (fchollet) [inline examples/cifar10_cnn.py]: No need to create a new dependency; this can be done with `os`.

File: examples/cifar10_cnn.py [hunk 3]
```diff
@@ -99,3 +109,37 @@
...
+
+# Save model and weights
+ensure_dir_exists(save_dir)
+model.save(save_dir+model_name)
+model.save_weights(save_dir+weights_name)
+print('Saved trained model: ' +save_dir+model_name)
+print('Saved weights: ' +save_dir+weights_name)
+
+# Load label names to use in prediction results
+label_list = os.path.expanduser("~") + '/.keras/datasets/cifar-10-batches-py/batches.meta'
```
Comment (fchollet) [inline examples/cifar10_cnn.py]: This will break for Windows users. Check how we construct file paths elsewhere (e.g. in `backend/common.py`).

File: examples/cifar10_cnn.py [hunk 4]
```diff
@@ -99,3 +109,37 @@
                         steps_per_epoch=x_train.shape[0] // batch_size,
                         epochs=epochs,
                         validation_data=(x_test, y_test))
+
+# Save model and weights
+ensure_dir_exists(save_dir)
+model.save(save_dir+model_name)
+model.save_weights(save_dir+weights_name)
+print('Saved trained model: ' +save_dir+model_name)
+print('Saved weights: ' +save_dir+weights_name)
```
Comment (fchollet) [inline examples/cifar10_cnn.py]: These few lines break PEP8 conventions.

File: examples/cifar10_cnn.py [hunk 5]
```diff
@@ -99,3 +109,37 @@
...
+label_list = os.path.expanduser("~") + '/.keras/datasets/cifar-10-batches-py/batches.meta'
+with open(label_list, mode='rb') as f:
+    labels = pickle.load(f)
+
+# Evaluate model with test data set and share sample prediction results
+evaluation = model.evaluate_generator(datagen.flow(x_test, y_test,
+                                      batch_size=batch_size),
+                                      steps=x_test.shape[0] // batch_size)
+
+eval_results = map(lambda metric, result: metric + ' = ' + str(result), \
```
Comment (fchollet) [inline examples/cifar10_cnn.py]: You don't need a lambda here. Also, don't break lines with `\`.

File: examples/cifar10_cnn.py [hunk 6]
```diff
@@ -99,3 +109,37 @@
...
+
+
+predict_y = model.predict_generator(datagen.flow(x_test, y_test,
+                                    batch_size=batch_size),
+                                    steps=x_test.shape[0] // batch_size)
+
+for idx, y in enumerate(predict_y):
+  print("Actual | Predicted = %s | %s"% (labels['label_names'][np.argmax(y_test[idx])], \
+                                          labels['label_names'][np.argmax(y)]))
+  if idx == num_pred_print: break
```
Comment (fchollet) [inline examples/cifar10_cnn.py]: Don't put logic on an `if` line, introduce a line break. In general, I suggest rewriting the part of the code that prints evaluation results to make it simpler, easier to read, and more user friendly. A big goal of these example scripts is to be as user friendly as possible while introducing best practices.

File: examples/cifar10_cnn.py (hunk 2 above)
Comment (nyghtowl) [inline examples/cifar10_cnn.py]: Thank you for the code review. I was using this dependency to simplify the call to check dir and make if it doesn't. I can switch to os and add the check in the example.

File: examples/cifar10_cnn.py (hunk 3 above)
Comment (nyghtowl) [inline examples/cifar10_cnn.py]: Didn't see an example in backend but does it work to follow what you have in data_utils? https://github.com/fchollet/keras/blob/master/keras/utils/data_utils.py#L171

**Comment by fchollet** [comment]
Actually, it looks like there are PEP8 issues left. Please fix.

https://travis-ci.org/fchollet/keras/jobs/257560704

File: examples/cifar10_cnn.py [hunk 7]
```diff
@@ -99,3 +108,45 @@
...
                         epochs=epochs,
                         validation_data=(x_test, y_test))
+
+# Save model and weights
+if not os.path.isdir(save_dir):
+    os.makedirs(save_dir)
+model_path = os.path.join(save_dir, model_name)
+weights_path = os.path.join(save_dir, weights_name)
+model.save(model_path)
+model.save_weights(weights_path)
```
Comment (fchollet) [inline examples/cifar10_cnn.py]: Didn't notice this earlier, but why save both the entire model and its weights, separately? The saved models already contains the weights. I suggest removing `save_weights`.

**Comment by dlpbc** [comment]
@nyghtowl , @fchollet 
The section of the cifar10_cnn.py code to evaluate the model and generate prediction fails to run correctly when 'data_augmentation' is set to False.

This is because 'datagen' variable (used to evaluate and generate model prediction) is defined only when 'data_augmentation' is set to True.
